File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prereq: List[List[int]]) -> bool:
        adj = [None]*numCourses
        for i in prereq:
            if adj[i[0]]:
                adj[i[0]].append(i[1])
            else:
                adj[i[0]]=[i[1]]
        # print(adj)
        
        vis1=[0]*numCourses
        vis2=[0]*numCourses
        
        
        def dfs(i):
            if vis1[i]!=0 and vis2[i]==0:
                return True
            if vis2[i]!=0:
                return False

            
            
            vis1[i]+=1
            vis2[i]+=1
            ans=True
            if adj[i]:
                for pre in adj[i]:
                    ans = ans and dfs(pre)
            vis2[i]-=1
            return ans
        
        
        
        for i in range(numCourses):
            if not dfs(i):
                return False
        return True
        #     if vis1[i]!=0:
        #         con
        #     if adj[i]==None:
        #         vis1[i]+=1
        #         # vis2[i]+=1
        #         continue
        
        
            

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int n, int[][] pre) {
        int[] indeg = new int[n];
        List<Integer>[] adj = new ArrayList[n];
        
        for(int i=0; i<n; ++i) adj[i] = new ArrayList<>();
        
        for(int[] e: pre){
            adj[e[0]].add(e[1]);
            ++indeg[e[1]];
        } 
        
        Queue<Integer> q = new LinkedList<>();
        int seen = 0;
        
        for(int i=0; i<n; ++i){
            if(indeg[i]==0) q.offer(i);
        }
        
        while(!q.isEmpty() && seen<n){
            int node = q.poll();
            ++seen;

            for(int nbr: adj[node]){
                --indeg[nbr];
                if(indeg[nbr]==0) q.offer(nbr);
            }
        }
        
        return seen==n;
    }
}



// // find sequence
//
// class Solution {
//     public boolean canFinish(int n, int[][] pre) {
//         Queue<Integer> q = new LinkedList<>();
//         List<Integer>[] adj = (List<Integer>[]) new List[n];
//         for (int i = 0; i < n; i++) {
//             adj[i] = new ArrayList<Integer>();
//         }

//         int[] inDeg = new int[n];
//         for(int[] p: pre){
//             adj[p[0]].add(p[1]);
//             inDeg[p[1]]+=1;
//         }
//         for(int i=0; i<n; ++i){
//             if(inDeg[i]==0)
//                 q.offer(i);
//         }
        
//         List<Integer> topo =  new ArrayList<>();
//         while(!q.isEmpty()){
//             int node = q.poll();
//             topo.add(node);
//             for(int nbr: adj[node]){
//                 inDeg[nbr]--;
//                 if(inDeg[nbr]==0)
//                     q.offer(nbr);
//             }
//         }
//         return topo.size()==n;
//     }
// }